File: lib/dynamic.go
/*
  Standard program for implementing dynamic programs 
  Copyright (c) 2016. Francis G. McCabe

  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
  except in compliance with the License. You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software distributed under the
  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, either express or implied. See the License for the specific language governing
  permissions and limitations under the License.
 */

go.dynamic{ 

  dynTest[T] <~ { check:[T]{} }.         -- We use this as a more sophisticated match

  dynamic[T] <~ { mem:[T]{}.		 -- Test for membership in the relation
	add:[T]*.			 -- Add element to relation
	del:[T]*.			 -- Delete element from relation
	delc:[dynTest[T]]*.	      -- Delete satisfying element from relation
	delall:[T]*.			 -- Delete all matching elements 
	delallc:[dynTest[T]]*.		 -- Delete all satisfying elements
	ext:[]=>list[T].		 -- pick up extension
	match:[dynTest[T]]=>list[T] }.	 -- pick up matching sub-relation

  dynamic:[list[T]]@>dynamic[T].
  dynamic(Init)..{
    Db:list[symbol] := initDB(Init).

    initDB:[list[t]]=>list[symbol].
    initDB([]) => [].
    initDB([E,..L]) => [__term(E),..initDB(L)].
    initDB(X) => raise error("problem with "<>X.show(),'fail').
       
    mem(P) :- Cl in Db, __is(Cl,P).
       
    add(N) -> sync{ Db := Db<>[__term(N)] }.
       
    del(E) -> sync{ append(F,[Cl,..R],Db), __is(Cl,E) ? Db := F<>R | {} }.

    delc(Tst) -> sync{ append(F,[Cl,..R],Db), __is(Cl,E),Tst.check(E) ? Db := F<>R | {} }.

    delall(E) -> sync{ Db := { Cl .. (Cl::\+__is(Cl,E)) in Db } }.

    delallc(Tst) -> sync{ Db := { Cl .. (Cl::\+(__is(Cl,E),Tst.check(E))) in Db }}.

    ext() => valof{
	       sync{
		 valis { E .. (Cl::__is(Cl,E)) in Db }
	       }
	     }.

    match(K) => valof{
		  sync{
		    valis { E .. (Cl::__is(Cl,E),K.check(E)) in Db }
		  }
		}.
    
    show()=> "{" <> showExt(ext(),"") <> "}".

    showExt:[list[symbol],string]=>string.
    showExt([],_)=>"".
    showExt([E,..Rl],Pre) => Pre<>E.show()<> showExt(Rl,". ").
  }.

}.

File: Engine/dynamic.c
/*
  Property management
  Copyright (c) 2016, 2017. Francis G. McCabe

  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
  except in compliance with the License. You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software distributed under the
  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, either express or implied. See the License for the specific language governing
  permissions and limitations under the License.
*/

#include <string.h>		/* Access string defs */

#include "lo.h"

ptrI dynamicClass;

static long dySizeFun(specialClassPo class, objPo o);
static comparison dyCompFun(specialClassPo class, objPo o1, objPo o2);
static retCode dyOutFun(specialClassPo class, ioPo out, objPo o);
static retCode dyScanFun(specialClassPo class, specialHelperFun helper, void *c, objPo o);
static objPo dyCopyFun(specialClassPo class, objPo dst, objPo src);
static uinteger dyHashFun(specialClassPo class, objPo o);

void initDynamicClass(void) {
  dynamicClass = newSpecialClass("#dynamic", dySizeFun, dyCompFun,
                                 dyOutFun, dyCopyFun, dyScanFun, dyHashFun);
}

static long dySizeFun(specialClassPo class, objPo o) {
  return CellCount(sizeof(dynRec));
}

static comparison dyCompFun(specialClassPo class, objPo o1, objPo o2) {
  if (o1 == o2)
    return same;
  else
    return incomparible;
}

static retCode dyOutFun(specialClassPo class, ioPo out, objPo o) {
  dynPo dObj = (dynPo) o;

  return outMsg(out, "obj#%ld", dObj->hash);
}

static retCode dyScanFun(specialClassPo class, specialHelperFun helper, void *c, objPo o) {
  assert(o->class == dynamicClass);

  dynPo d = (dynPo) o;

  return helper(&d->code, c);
}

static objPo dyCopyFun(specialClassPo class, objPo dst, objPo src) {
  long size = dySizeFun(class, src);
  memmove((void *) dst, (void *) src, size * sizeof(ptrI));

  return (objPo) (((ptrPo) dst) + size);
}

static uinteger dyHashFun(specialClassPo class, objPo o) {
  assert(o->class == dynamicClass);

  dynPo d = (dynPo) o;

  return d->hash;
}

static uinteger objNumber = 0;

ptrI dynamicObject(heapPo H) {
  dynPo perm = (dynPo) permAllocate(CellCount(sizeof(dynRec)));

  perm->class = dynamicClass;
  perm->code = kvoid;
  perm->lock = newLock();
  perm->hash = objNumber++;

  return objP(perm);
}

void setDynamicCode(ptrI O, ptrI code) {
  assert(hasClass(objV(O), dynamicClass));
  dynPo o = (dynPo) objV(O);
  o->code = code;
}
